src/server/main_service.cc

```diff
@@ -1326,6 +1325,123 @@ bool StartMultiExec(DbIndex dbid, Transaction* trans, ConnectionState::ExecInfo*
   return true;
 }
 
+struct ExecutionSquasher {
+  static void Execute(absl::Span<StoredCmd> cmds, ConnectionContext* cntx) {
+    ExecutionSquasher es{cmds, cntx};
+    es.Iterate();
+  }
+
+ private:
+  ExecutionSquasher(absl::Span<StoredCmd> cmds, ConnectionContext* cntx)
+      : cmds_{cmds}, cntx_{cntx}, base_cid_{cntx->transaction->GetCId()} {
+  }
+
+  bool CheckSquashing(StoredCmd* cmd) {
+    auto args = cmd->ArgList();
+    auto keys = DetermineKeys(cmd->descr, args);
+    if (!keys.ok())
+      return false;
+
+    // Count number of shards.
+    tmp_shards_.clear();
+    for (unsigned i = keys->start; i < keys->end; i += keys->step)
+      tmp_shards_.insert(Shard(ArgS(args, i), shard_set->size()));
+
+    if (keys->bonus)
+      tmp_shards_.insert(Shard(ArgS(args, keys->bonus), shard_set->size()));
+
+    if (tmp_shards_.size() > 1)
+      return false;
+
+    if (sharded_.empty())
+      sharded_.resize(shard_set->size());
+
+    ShardId sid = *tmp_shards_.begin();
+    order_.push_back(sid);
+    sharded_[sid].cmds.push_back(cmd);
+    return true;
+  }
+
+  void ExecuteOne(StoredCmd* cmd) {
+    DCHECK(order_.empty());
+    cntx_->transaction->MultiSwitchCmd(cmd->descr);
+    if (IsTransactional(cmd->descr))
+      cntx_->transaction->InitByArgs(cntx_->conn_state.db_index, cmd->ArgList());
+    cmd->descr->Invoke(cmd->ArgList(), cntx_);
+  }
+
+  OpStatus ShardCb(Transaction* parent_tx, EngineShard* es) {
+    auto& sinfo = sharded_[es->shard_id()];
+    if (sinfo.cmds.empty())
+      return OpStatus::OK;
+
+    intrusive_ptr<Transaction> local_tx{new Transaction{parent_tx}};
+    ConnectionContext local_cntx{local_tx.get()};
+
+    for (auto* cmd : sinfo.cmds) {
+      local_tx->MultiSwitchCmd(cmd->descr);
+      local_tx->InitByArgs(parent_tx->GetDbIndex(), cmd->ArgList());
+      cmd->descr->Invoke(cmd->ArgList(), &local_cntx);
+
+      auto reply = local_cntx.GetCapture();
+      sinfo.replies.emplace_back(move(reply));
+    }
+
+    return OpStatus::OK;
+  }
+
+  void ExecuteSquashed() {
```

Comment:
I understand this file what creates the magic. Please add some explanation about the algo in the commit/pr description,